File: 29-04-2025/4_Median_of_Two_Sorted_Arrays/MergeSort_O()/solution1.js
// https://leetcode.com/problems/median-of-two-sorted-arrays/?envType=problem-list-v2&envId=array

/*
* Given two sorted arrays nums1 and nums2 of size m and n respectively,
* return the median of the two sorted arrays.
*
* The overall run time complexity should be O(log (m+n)).
*
Example 1:

Input: nums1 = [1,3], nums2 = [2]
Output: 2.00000
Explanation: merged array = [1,2,3] and median is 2.
Example 2:

Input: nums1 = [1,2], nums2 = [3,4]
Output: 2.50000
Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.

* */

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
var findMedianSortedArrays = function (nums1, nums2) {
    let m = nums1.length,
        n = nums2.length;
    let p1 = 0,
        p2 = 0;

    let getMin = function () {
        console.log(p1, p2)
        if (p1 < m && p2 < n) {
            return nums1[p1] < nums2[p2] ? nums1[p1++] : nums2[p2++];
        } else if (p1 < m) {
            return nums1[p1++];
        } else if (p2 < n) {
            return nums2[p2++];
        }
        return -1;
    };

    if ((m + n) % 2 === 0) {
        console.log('odd', (m + n) / 2 - 1);
        for (let i = 0; i < (m + n) / 2 - 1; ++i) {
            let temp = getMin();
        }
        return (getMin() + getMin()) / 2.0;
    } else {
        for (let i = 0; i < Math.floor((m + n) / 2); ++i) {
            let temp = getMin();
        }
        return getMin();
    }
};

// console.log(findMedianSortedArrays([1, 3], [2])); // 2.00000
console.log(findMedianSortedArrays([1,2], [3,4])); // 2.50000

// console.log(findMedianSortedArrays([], [1]));

// console.log(findMedianSortedArrays([1,2,3,4,5], [6,7,8,9,10,11,12,13,14,15,16,17]));

File: 04-05-2025/27_Remove_Element/solutions/src/solution1.java
//https://leetcode.com/problems/remove-element/submissions/1623423457/?envType=problem-list-v2&envId=array

import java.util.Arrays;

class solution1 {
    public static void main(String[] args) {
        solution1 solution1 = new solution1();

        // 테스트 케이스 1
        int[] nums1 = {3, 2, 2, 3};
        int val1 = 3;
        int k1 = solution1.removeElement(nums1, val1);
        System.out.println("Input: nums = " + Arrays.toString(nums1) + ", val = " + val1);
        System.out.println("Output: k = " + k1 + ", nums = " + Arrays.toString(Arrays.copyOf(nums1, k1))); // 예상 출력: k = 2, nums = [2, 2]

        // 테스트 케이스 2
        int[] nums2 = {0, 1, 2, 2, 3, 0, 4, 2};
        int val2 = 2;
        int k2 = solution1.removeElement(nums2, val2);
        System.out.println("Input: nums = " + Arrays.toString(nums2) + ", val = " + val2);
        System.out.println("Output: k = " + k2 + ", nums = " + Arrays.toString(Arrays.copyOf(nums2, k2))); // 예상 출력: k = 5, nums = [0, 1, 3, 0, 4]
    }

    public int removeElement(int[] nums, int val) {
        int n = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != val) {
                nums[n] = nums[i];
                n += 1;
            }
        }
        return n;
    }
}